{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/AlexanderCoudijzer/BM25-VSM-Search-Engine/blob/main/Search_engine_v0_0.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XeS4WON90flx"
      },
      "source": [
        "# Data loading testing\n",
        "see https://ir-datasets.com/beir.html#beir/scifact  \n",
        "and https://ir-datasets.com/python.html"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8CvP86nnwTuw"
      },
      "source": [
        "##Data Loading"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "5Qh8yMNt6QeI"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "!pip install --upgrade ir_datasets\n",
        "\n",
        "import pandas as pd\n",
        "from sklearn.feature_extraction.text import CountVectorizer\n",
        "import numpy as np\n",
        "import ir_datasets"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "i1g2Gf2LjToC",
        "outputId": "4c07649c-8909-4afc-b591-2948f914d5bc"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "5183 documents with: OrderedDict([('doc_id', <class 'str'>), ('text', <class 'str'>), ('title', <class 'str'>)])\n",
            "809 queries with: OrderedDict([('query_id', <class 'str'>), ('text', <class 'str'>)])\n",
            "919 qrels with: OrderedDict([('query_id', <class 'str'>), ('doc_id', <class 'str'>), ('relevance', <class 'int'>), ('iteration', <class 'str'>)])\n"
          ]
        }
      ],
      "source": [
        "dataset = ir_datasets.load(\"beir/scifact/train\")\n",
        "\n",
        "print(dataset.docs_count(), 'documents with:', dataset.docs_cls().__annotations__)\n",
        "print(dataset.queries_count(), 'queries with:', dataset.queries_cls().__annotations__)\n",
        "print(dataset.qrels_count(), 'qrels with:',dataset.qrels_cls().__annotations__)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "WIDi1JwIjh7R"
      },
      "outputs": [],
      "source": [
        "%%capture\n",
        "#this actually downloads the data, muted to avoid clutter\n",
        "for doc in dataset.docs_iter()[:1]: \n",
        "    None\n",
        "for q in dataset.queries_iter():\n",
        "    None\n",
        "for qrel in dataset.qrels_iter():\n",
        "    None"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "pkhFTCerkCNh"
      },
      "outputs": [],
      "source": [
        "corpus=[]\n",
        "for doc in dataset.docs_iter():\n",
        "    corpus.append([doc[0],doc[1],doc[2]])\n",
        "\n",
        "documents = [doc[1] for doc in corpus] # only taking the text for simplicity, but we'll need the doc_id for the evaluation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1YxFIFKTBNLM"
      },
      "source": [
        "## Scifact initial test"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "G3E7sQ5_kg9x",
        "outputId": "de33d7cd-7dbb-46bb-97dd-1b68bfbc6f3d"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n",
              "0      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "1      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "2      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "3      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "4      0    0     0      0      0       0      0      0      0      0  ...   \n",
              "...   ..  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n",
              "5178   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5179   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5180   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5181   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "5182   0    0     0      0      0       0      0      0      0      0  ...   \n",
              "\n",
              "      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin  zz  zzw  \\\n",
              "0            0       0        0        0           0        0      0   0    0   \n",
              "1            0       0        0        0           0        0      0   0    0   \n",
              "2            0       0        0        0           0        0      0   0    0   \n",
              "3            0       0        0        0           0        0      0   0    0   \n",
              "4            0       0        0        0           0        0      0   0    0   \n",
              "...        ...     ...      ...      ...         ...      ...    ...  ..  ...   \n",
              "5178         0       0        0        0           0        0      0   0    0   \n",
              "5179         0       0        0        0           0        0      0   0    0   \n",
              "5180         0       0        0        0           0        0      0   0    0   \n",
              "5181         0       0        0        0           0        0      0   0    0   \n",
              "5182         0       0        0        0           0        0      0   0    0   \n",
              "\n",
              "      zzz  \n",
              "0       0  \n",
              "1       0  \n",
              "2       0  \n",
              "3       0  \n",
              "4       0  \n",
              "...   ...  \n",
              "5178    0  \n",
              "5179    0  \n",
              "5180    0  \n",
              "5181    0  \n",
              "5182    0  \n",
              "\n",
              "[5183 rows x 35568 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-36797e3e-e562-4d60-b730-af09c3ca5e9d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0000</th>\n",
              "      <th>00001</th>\n",
              "      <th>00002</th>\n",
              "      <th>000035</th>\n",
              "      <th>00004</th>\n",
              "      <th>00005</th>\n",
              "      <th>00006</th>\n",
              "      <th>00008</th>\n",
              "      <th>...</th>\n",
              "      <th>zygosity</th>\n",
              "      <th>zygote</th>\n",
              "      <th>zygotes</th>\n",
              "      <th>zygotic</th>\n",
              "      <th>zymography</th>\n",
              "      <th>zymosan</th>\n",
              "      <th>zyxin</th>\n",
              "      <th>zz</th>\n",
              "      <th>zzw</th>\n",
              "      <th>zzz</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5178</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5179</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5180</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5181</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5182</th>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5183 rows × 35568 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-36797e3e-e562-4d60-b730-af09c3ca5e9d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-36797e3e-e562-4d60-b730-af09c3ca5e9d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-36797e3e-e562-4d60-b730-af09c3ca5e9d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ],
      "source": [
        "vectorizer = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "documents_vectorized = vectorizer.fit_transform(documents)\n",
        "vocabulary = vectorizer.get_feature_names_out()\n",
        "dataframe = pd.DataFrame(documents_vectorized.toarray(), columns=vocabulary)\n",
        "dataframe"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "OtpbYAsEk7up",
        "outputId": "faabd124-02ae-4805-e925-cc796fc5f98f"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "       00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n",
              "0     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "1     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "2     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "3     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "4     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "...   ...  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n",
              "5178  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5179  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5180  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5181  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "5182  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n",
              "\n",
              "      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin   zz  \\\n",
              "0          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "1          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "2          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "3          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "4          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "...        ...     ...      ...      ...         ...      ...    ...  ...   \n",
              "5178       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5179       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5180       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5181       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "5182       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n",
              "\n",
              "      zzw  zzz  \n",
              "0     0.0  0.0  \n",
              "1     0.0  0.0  \n",
              "2     0.0  0.0  \n",
              "3     0.0  0.0  \n",
              "4     0.0  0.0  \n",
              "...   ...  ...  \n",
              "5178  0.0  0.0  \n",
              "5179  0.0  0.0  \n",
              "5180  0.0  0.0  \n",
              "5181  0.0  0.0  \n",
              "5182  0.0  0.0  \n",
              "\n",
              "[5183 rows x 35568 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-d779d1f1-ba08-4f91-9062-c549d7cb4df0\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>00</th>\n",
              "      <th>000</th>\n",
              "      <th>0000</th>\n",
              "      <th>00001</th>\n",
              "      <th>00002</th>\n",
              "      <th>000035</th>\n",
              "      <th>00004</th>\n",
              "      <th>00005</th>\n",
              "      <th>00006</th>\n",
              "      <th>00008</th>\n",
              "      <th>...</th>\n",
              "      <th>zygosity</th>\n",
              "      <th>zygote</th>\n",
              "      <th>zygotes</th>\n",
              "      <th>zygotic</th>\n",
              "      <th>zymography</th>\n",
              "      <th>zymosan</th>\n",
              "      <th>zyxin</th>\n",
              "      <th>zz</th>\n",
              "      <th>zzw</th>\n",
              "      <th>zzz</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5178</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5179</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5180</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5181</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5182</th>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5183 rows × 35568 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d779d1f1-ba08-4f91-9062-c549d7cb4df0')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-d779d1f1-ba08-4f91-9062-c549d7cb4df0 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-d779d1f1-ba08-4f91-9062-c549d7cb4df0');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ],
      "source": [
        "# Defining all the variables we need:\n",
        "k_1 = 1.2\n",
        "b = 0.8\n",
        "dfs = (dataframe > 0).sum(axis=0) # doc frequency\n",
        "N = dataframe.shape[0] # total number of docs\n",
        "idfs = np.log10(N/dfs) # inverse doc frequency\n",
        "#dls = [len(d.split(' ')) for d in documents] # considering all words in doc\n",
        "dls = dataframe.sum(axis=1).tolist() # considering words minus stop words in doc (better option)\n",
        "avgdl = np.mean(dls) # single value, mean doc length (minus stop words)\n",
        "\n",
        "# Applying the BM25 formula:\n",
        "numerator = np.array((k_1 + 1) * dataframe)\n",
        "denominator = np.array(k_1 *((1 - b) + b * (dls / avgdl))).reshape(N,1) + np.array(dataframe)\n",
        "BM25_tf = numerator / denominator\n",
        "idfs = np.array(idfs)\n",
        "BM25_score = BM25_tf * idfs\n",
        "\n",
        "bm25_idf = pd.DataFrame(BM25_score, columns=vocabulary)\n",
        "bm25_idf"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PotSpiIxltBy",
        "outputId": "66fb0aba-747f-4962-9bc2-e58c5ad7ed3d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['1 in 5 million in UK have abnormal PrP positivity.']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['abnormal', 'million', 'positivity', 'prp', 'uk'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ],
      "source": [
        "query_num = 1 #query number to test\n",
        "queries = []\n",
        "for q in dataset.queries_iter(): # reading in all the queries\n",
        "    queries.append([q[0],q[1]])\n",
        "print([queries[query_num][1]])\n",
        "\n",
        "vectorizer_q = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "query_vectorized = vectorizer_q.fit_transform([queries[query_num][1]])\n",
        "q_terms = vectorizer_q.get_feature_names_out()\n",
        "q_terms"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ypvq9CGBl6Is",
        "outputId": "f5064a8f-8453-447c-c42e-33041b5d9f6e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(5183, 5)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('OBJECTIVES To carry out a further survey of archived appendix samples to understand better the differences between existing estimates of the prevalence of subclinical infection with prions after the bovine spongiform encephalopathy epizootic and to see whether a broader birth cohort was affected, and to understand better the implications for the management of blood and blood products and for the handling of surgical instruments. DESIGN Irreversibly unlinked and anonymised large scale survey of archived appendix samples. SETTING Archived appendix samples from the pathology departments of 41 UK hospitals participating in the earlier survey, and additional hospitals in regions with lower levels of participation in that survey. SAMPLE 32,441 archived appendix samples fixed in formalin and embedded in paraffin and tested for the presence of abnormal prion protein (PrP). RESULTS Of the 32,441 appendix samples 16 were positive for abnormal PrP, indicating an overall prevalence of 493 per million population (95% confidence interval 282 to 801 per million). The prevalence in those born in 1941-60 (733 per million, 269 to 1596 per million) did not differ significantly from those born between 1961 and 1985 (412 per million, 198 to 758 per million) and was similar in both sexes and across the three broad geographical areas sampled. Genetic testing of the positive specimens for the genotype at PRNP codon 129 revealed a high proportion that were valine homozygous compared with the frequency in the normal population, and in stark contrast with confirmed clinical cases of vCJD, all of which were methionine homozygous at PRNP codon 129. CONCLUSIONS This study corroborates previous studies and suggests a high prevalence of infection with abnormal PrP, indicating vCJD carrier status in the population compared with the 177 vCJD cases to date. These findings have important implications for the management of blood and blood products and for the handling of surgical instruments.',\n",
              "  11.537320382799532),\n",
              " ('We report a case of preclinical variant Creutzfeldt-Jakob disease (vCJD) in a patient who died from a non-neurological disorder 5 years after receiving a blood transfusion from a donor who subsequently developed vCJD. Protease-resistant prion protein (PrP(res)) was detected by western blot, paraffin-embedded tissue blot, and immunohistochemistry in the spleen, but not in the brain. Immunohistochemistry for prion protein was also positive in a cervical lymph node. The patient was a heterozygote at codon 129 of PRNP, suggesting that susceptibility to vCJD infection is not confined to the methionine homozygous PRNP genotype. These findings have major implications for future estimates and surveillance of vCJD in the UK.',\n",
              "  6.3384804894944935),\n",
              " (\"The prion protein (PrP(C)) is highly expressed in the nervous system and critically involved in prion diseases where it misfolds into pathogenic PrP(Sc). Moreover, it has been suggested as a receptor mediating neurotoxicity in common neurodegenerative proteinopathies such as Alzheimer's disease. PrP(C) is shed at the plasma membrane by the metalloprotease ADAM10, yet the impact of this on prion disease remains enigmatic. Employing conditional knockout mice, we show that depletion of ADAM10 in forebrain neurons leads to posttranslational increase of PrP(C) levels. Upon prion infection of these mice, clinical, biochemical, and morphological data reveal that lack of ADAM10 significantly reduces incubation times and increases PrP(Sc) formation. In contrast, spatiotemporal analysis indicates that absence of shedding impairs spread of prion pathology. Our data support a dual role for ADAM10-mediated shedding and highlight the role of proteolytic processing in prion disease.\",\n",
              "  5.479213806892906),\n",
              " ('Native mammalian prions exist in self-propagating strains that exhibit distinctive clinical, pathological and biochemical characteristics. Prion strain diversity is associated with variations in PrP(Sc) conformation, but it remains unknown precisely which physical properties of the PrP(Sc) molecules are required to encipher mammalian prion strain phenotypes. In this study, we subjected prion-infected brain homogenates derived from three different hamster scrapie strains to either (i) proteinase K digestion or (ii) sonication, and inoculated the modified samples into normal hamsters. The results show that the strain-specific clinical features and neuropathological profiles of inoculated animals were not affected by either treatment. Similarly, the strain-dependent biochemical characteristics of the PrP(Sc) molecules (including electrophoretic mobility, glycoform composition, conformational stability and susceptibility to protease digestion) in infected animals were unaffected by either proteolysis or sonication of the original inocula. These results indicate that the infectious strain properties of native prions do not appear to be altered by PrP(Sc) disaggregation, and that maintenance of such properties does not require the N-domain (approximately residues 23-90) of the protease-resistant PrP(Sc) molecules or protease-sensitive PrP(Sc) molecules.',\n",
              "  5.42241988394522),\n",
              " ('Prions are the protein-based infectious agents responsible for prion diseases. Environmental prion contamination has been implicated in disease transmission. Here, we analyzed the binding and retention of infectious prion protein (PrP(Sc)) to plants. Small quantities of PrP(Sc) contained in diluted brain homogenate or in excretory materials (urine and feces) can bind to wheat grass roots and leaves. Wild-type hamsters were efficiently infected by ingestion of prion-contaminated plants. The prion-plant interaction occurs with prions from diverse origins, including chronic wasting disease. Furthermore, leaves contaminated by spraying with a prion-containing preparation retained PrP(Sc) for several weeks in the living plant. Finally, plants can uptake prions from contaminated soil and transport them to aerial parts of the plant (stem and leaves). These findings demonstrate that plants can efficiently bind infectious prions and act as carriers of infectivity, suggesting a possible role of environmental prion contamination in the horizontal transmission of the disease.',\n",
              "  4.827903110869903),\n",
              " ('OBJECTIVE To assess the robustness of patient responses to a new national survey of patient experience as a basis for providing financial incentives to doctors. DESIGN Analysis of the representativeness of the respondents to the GP Patient Survey compared with those who were sampled (5.5 million patients registered with 8273 general practices in England in January 2009) and with the general population. Analysis of non-response bias looked at the relation between practice response rates and scores on the survey. Analysis of the reliability of the survey estimated the proportion of the variance of practice scores attributable to true differences between practices. RESULTS The overall response rate was 38.2% (2.2 million responses), which is comparable to that in surveys using similar methodology in the UK. Men, young adults, and people living in deprived areas were under-represented among respondents. However, for questions related to pay for performance, there was no systematic association between response rates and questionnaire scores. Two questions which triggered payments to general practitioners were reliable measures of practice performance, with average practice-level reliability coefficients of 93.2% and 95.0%. Less than 3% and 0.5% of practices had fewer than the number of responses required to achieve conventional reliability levels of 90% and 70%. A change to the payment formula in 2009 resulted in an increase in the average impact of random variation in patient scores on payments to general practitioners compared with payments made in 2007 and 2008. CONCLUSIONS There is little evidence to support the concern of some general practitioners that low response rates and selective non-response bias have led to systematic unfairness in payments attached to questionnaire scores. The study raises issues relating to the validity and reliability of payments based on patient surveys and provides lessons for the UK and for other countries considering the use of patient experience as part of pay for performance schemes.',\n",
              "  4.802553037383037),\n",
              " ('Infectious prions propagate from peripheral entry sites into the central nervous system (CNS), where they cause progressive neurodegeneration that ultimately leads to death. Yet the pathogenesis of prion disease can vary dramatically depending on the strain, or conformational variant of the aberrantly folded and aggregated protein, PrP(Sc). Although most prion strains invade the CNS, some prion strains cannot gain entry and do not cause clinical signs of disease. The conformational basis for this remarkable variation in the pathogenesis among strains is unclear. Using mouse-adapted prion strains, here we show that highly neuroinvasive prion strains primarily form diffuse aggregates in brain and are noncongophilic, conformationally unstable in denaturing conditions, and lead to rapidly lethal disease. These neuroinvasive strains efficiently generate PrP(Sc) over short incubation periods. In contrast, the weakly neuroinvasive prion strains form large fibrillary plaques and are stable, congophilic, and inefficiently generate PrP(Sc) over long incubation periods. Overall, these results indicate that the most neuroinvasive prion strains are also the least stable, and support the concept that the efficient replication and unstable nature of the most rapidly converting prions may be a feature linked to their efficient spread into the CNS.',\n",
              "  4.688771720724594),\n",
              " ('Hemolysates of erythrocytes from more than a quarter million people in Alabama were electrophoresed on cellulose acetate, pH 8.4, and those samples exhibiting an abnormality were also electrophoresed in citrate agar, pH 6.0. The globin chains of mutants other than Hb S and C were electrophoresed in urea-mercaptoethanol buffers at both pH 8.9 and pH 6.0, and 60 of them were also analyzed structurally. Of about 6000 samples from whites, only three contained abnormal hemoglobins--Hb D Los Angeles, Hb J Baltimore, and one unidentified. Of 249,000 samples from blacks, about 29,000 contained electrophoretically detectable abnormalities, most of them associated with Hb S or C, present in a frequency of about 9% and 3%, respectively. About 1000 samples resolved into patterns of potential clinical significance. Twenty other mutant hemoglobins were detected, in various genetic combinations in 164 kindreds; four of these-Hb Alabama, Montgomery, Titusville, and Mobile--were previously unknown. The methods used are rapid, economical, and well suited for large scale surveys. They provide highly specific characterizations of many mutant hemoglobins, and no discrepancies were found between the presumptive identifications based on these characterizations and the definitive identifications obtained from structural analyses.',\n",
              "  3.8640302304421676),\n",
              " ('A considerable subgroup of patients with early breast cancer does not address benefits of anthracycline based chemotherapy. The aim of this retrospective study was to investigate the effect of microvessel density (MVD) and status of p53 protein on 5-year disease free survival (DFS) in the group of breast cancer patients treated with anthracyclines in adjuvant setting. Correlations between MVD, p53 status and other clinicopathological parameters were also assessed. MVD and p53 status were analyzed immunohistochemically in the group of 172 women with breast cancer in clinical stage T1-2, N1-N2, M0. There were 123 tumors (71.5 %) with lower MVD (≤ 214.8 microvesells/mm(2)) and 49 (28.5 %) with higher MVD (>214.8 microvesells/mm(2)). The proportion of higher MVD tumors significantly increased in N2 (P = 0.000) and in estrogen (P = 0.046) or progesterone receptors (P = 0.029) negative tumors. p53 positivity was indicated in 50 cancers (29.1 %) and was significantly associated with higher grade (P = 0.000), steroid receptors negativity (P = 0.000), cytokeratin5/6 positivity (P = 0.026), topoisomerase IIα overexpression (P = 0.005) and higher proliferation rate (P = 0.001). In univariate analysis, higher MVD (P = 0.016) and p53 negativity (P = 0.023) were significantly related with longer DFS (median follow-up 36 months). In multivariate Cox regression analysis MVD was independently associated with DFS. These data suggest that higher MVD is favourable prognostic factors for early advanced breast cancer patients after adjuvant anthracycline based chemotherapy.',\n",
              "  3.6448981560117626),\n",
              " ('OBJECTIVE To determine whether myelin oligodendrocyte glycoprotein antibodies (MOG-Abs) were predictive of a demyelination phenotype in children presenting with acquired demyelinating syndrome (ADS). METHOD Sixty-five children with a first episode of ADS (12 acute disseminated encephalomyelitis, 24 optic neuritis, 18 transverse myelitis, 11 other clinically isolated syndrome) were identified from 2 national demyelination programs in the United Kingdom and France. Acute serum samples were tested for MOG-Abs by cell-based assay. Antibodies were used to predict diagnosis of multiple sclerosis (MS) at 1 year. RESULTS Twenty-three of 65 (35%) children had MOG-Abs. Antibody-positive and antibody-negative patients were not clinically different at presentation, but identification of MOG-Abs predicted a non-MS course at 1-year follow-up: only 2/23 (9%) MOG-Ab-positive patients were diagnosed with MS compared to 16/42 (38%) MOG-Ab-negative patients (p = 0.019, Fisher exact test). Antibody positivity at outset was a useful predictor for a non-MS disease course, with a positive predictive value of 91% (95% confidence interval [CI] 72-99), negative predictive value of 38% (95% CI 24-54), positive likelihood ratio of 4.02 (CI 1.0-15.4), and odds ratio of 6.5 (CI 1.3-31.3). CONCLUSIONS MOG-Abs are found at presentation in 35% of patients with childhood ADS, across a range of demyelinating disorders. Antibody positivity can be useful in predicting a non-MS disease course at onset.',\n",
              "  3.621069192264962)]"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ],
      "source": [
        "q_terms = [term for term in q_terms if term in bm25_idf.columns]\n",
        "q_terms_only_df = bm25_idf[q_terms]\n",
        "score_q_d = q_terms_only_df.sum(axis=1)\n",
        "print(q_terms_only_df.shape)\n",
        "sorted(zip(documents,score_q_d.values), key = lambda tup:tup[1], reverse=True)[:10]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "EH1lTrrXcKy0"
      },
      "outputs": [],
      "source": [
        "qrels = []\n",
        "for qrel in dataset.qrels_iter(): # reading in all the relevancy scores\n",
        "    qrels.append(qrel)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7OvjLQnH80JX"
      },
      "source": [
        "## VSM"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "9awxsDRA8zAR"
      },
      "outputs": [],
      "source": [
        "dfs = (dataframe > 0).sum(axis=0) # doc frequency\n",
        "N = dataframe.shape[0] # total number of docs\n",
        "idfs = np.log10(N/dfs) # inverse doc frequency\n",
        "tfidf_frame = dataframe * idfs"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9swJUzKwg1sF",
        "outputId": "d4fc96eb-16e9-4234-a0be-9552a87d2b69"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['1 in 5 million in UK have abnormal PrP positivity.']\n"
          ]
        }
      ],
      "source": [
        "query_num = 1 #query number to test\n",
        "queries = []\n",
        "for q in dataset.queries_iter(): # reading in all the queries\n",
        "    queries.append([q[0],q[1]])\n",
        "print([queries[query_num][1]])\n",
        "q_vectorizer = CountVectorizer(stop_words='english', strip_accents='ascii')\n",
        "q = q_vectorizer.fit_transform([queries[query_num][1]])\n",
        "q_voc = q_vectorizer.get_feature_names_out()\n",
        "\n",
        "q_vector =[]\n",
        "for t in dataframe.columns:\n",
        "    if t in q_voc: q_vector.append(1)\n",
        "    else: q_vector.append(0)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "doc_vector = np.array(tfidf_frame)\n",
        "q_vector = np.array(q_vector)\n",
        "#doc_vector.shape, q_vector.shape\n"
      ],
      "metadata": {
        "id": "fWOe75cjBEu9"
      },
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from numpy.linalg import norm\n",
        "VSM_scores = np.dot(doc_vector, q_vector)/(norm(doc_vector)*norm(q_vector))\n",
        "sorted(zip(documents,VSM_scores), key = lambda tup:tup[1], reverse=True)[:10]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kgfExxyZHf3H",
        "outputId": "f8c50c97-0a9e-490b-f878-1c251bed38fb"
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('BACKGROUND The Global Burden of Diseases, Injuries, and Risk Factors Study 2015 provides an up-to-date synthesis of the evidence for risk factor exposure and the attributable burden of disease. By providing national and subnational assessments spanning the past 25 years, this study can inform debates on the importance of addressing risks in context. METHODS We used the comparative risk assessment framework developed for previous iterations of the Global Burden of Disease Study to estimate attributable deaths, disability-adjusted life-years (DALYs), and trends in exposure by age group, sex, year, and geography for 79 behavioural, environmental and occupational, and metabolic risks or clusters of risks from 1990 to 2015. This study included 388 risk-outcome pairs that met World Cancer Research Fund-defined criteria for convincing or probable evidence. We extracted relative risk and exposure estimates from randomised controlled trials, cohorts, pooled cohorts, household surveys, census data, satellite data, and other sources. We used statistical models to pool data, adjust for bias, and incorporate covariates. We developed a metric that allows comparisons of exposure across risk factors-the summary exposure value. Using the counterfactual scenario of theoretical minimum risk level, we estimated the portion of deaths and DALYs that could be attributed to a given risk. We decomposed trends in attributable burden into contributions from population growth, population age structure, risk exposure, and risk-deleted cause-specific DALY rates. We characterised risk exposure in relation to a Socio-demographic Index (SDI). FINDINGS Between 1990 and 2015, global exposure to unsafe sanitation, household air pollution, childhood underweight, childhood stunting, and smoking each decreased by more than 25%. Global exposure for several occupational risks, high body-mass index (BMI), and drug use increased by more than 25% over the same period. All risks jointly evaluated in 2015 accounted for 57·8% (95% CI 56·6-58·8) of global deaths and 41·2% (39·8-42·8) of DALYs. In 2015, the ten largest contributors to global DALYs among Level 3 risks were high systolic blood pressure (211·8 million [192·7 million to 231·1 million] global DALYs), smoking (148·6 million [134·2 million to 163·1 million]), high fasting plasma glucose (143·1 million [125·1 million to 163·5 million]), high BMI (120·1 million [83·8 million to 158·4 million]), childhood undernutrition (113·3 million [103·9 million to 123·4 million]), ambient particulate matter (103·1 million [90·8 million to 115·1 million]), high total cholesterol (88·7 million [74·6 million to 105·7 million]), household air pollution (85·6 million [66·7 million to 106·1 million]), alcohol use (85·0 million [77·2 million to 93·0 million]), and diets high in sodium (83·0 million [49·3 million to 127·5 million]). From 1990 to 2015, attributable DALYs declined for micronutrient deficiencies, childhood undernutrition, unsafe sanitation and water, and household air pollution; reductions in risk-deleted DALY rates rather than reductions in exposure drove these declines. Rising exposure contributed to notable increases in attributable DALYs from high BMI, high fasting plasma glucose, occupational carcinogens, and drug use. Environmental risks and childhood undernutrition declined steadily with SDI; low physical activity, high BMI, and high fasting plasma glucose increased with SDI. In 119 countries, metabolic risks, such as high BMI and fasting plasma glucose, contributed the most attributable DALYs in 2015. Regionally, smoking still ranked among the leading five risk factors for attributable DALYs in 109 countries; childhood underweight and unsafe sex remained primary drivers of early death and disability in much of sub-Saharan Africa. INTERPRETATION Declines in some key environmental risks have contributed to declines in critical infectious diseases. Some risks appear to be invariant to SDI. Increasing risks, including high BMI, high fasting plasma glucose, drug use, and some occupational exposures, contribute to rising burden from some conditions, but also provide opportunities for intervention. Some highly preventable risks, such as smoking, remain major causes of attributable DALYs, even as exposure is declining. Public policy makers need to pay attention to the risks that are increasingly major contributors to global burden. FUNDING Bill & Melinda Gates Foundation.',\n",
              "  0.009078057537868153),\n",
              " ('OBJECTIVES To carry out a further survey of archived appendix samples to understand better the differences between existing estimates of the prevalence of subclinical infection with prions after the bovine spongiform encephalopathy epizootic and to see whether a broader birth cohort was affected, and to understand better the implications for the management of blood and blood products and for the handling of surgical instruments. DESIGN Irreversibly unlinked and anonymised large scale survey of archived appendix samples. SETTING Archived appendix samples from the pathology departments of 41 UK hospitals participating in the earlier survey, and additional hospitals in regions with lower levels of participation in that survey. SAMPLE 32,441 archived appendix samples fixed in formalin and embedded in paraffin and tested for the presence of abnormal prion protein (PrP). RESULTS Of the 32,441 appendix samples 16 were positive for abnormal PrP, indicating an overall prevalence of 493 per million population (95% confidence interval 282 to 801 per million). The prevalence in those born in 1941-60 (733 per million, 269 to 1596 per million) did not differ significantly from those born between 1961 and 1985 (412 per million, 198 to 758 per million) and was similar in both sexes and across the three broad geographical areas sampled. Genetic testing of the positive specimens for the genotype at PRNP codon 129 revealed a high proportion that were valine homozygous compared with the frequency in the normal population, and in stark contrast with confirmed clinical cases of vCJD, all of which were methionine homozygous at PRNP codon 129. CONCLUSIONS This study corroborates previous studies and suggests a high prevalence of infection with abnormal PrP, indicating vCJD carrier status in the population compared with the 177 vCJD cases to date. These findings have important implications for the management of blood and blood products and for the handling of surgical instruments.',\n",
              "  0.004433917243939952),\n",
              " ('BACKGROUND Dengue is the most common arbovirus infection globally, but its burden is poorly quantified. We estimated dengue mortality, incidence, and burden for the Global Burden of Disease Study 2013. METHODS We modelled mortality from vital registration, verbal autopsy, and surveillance data using the Cause of Death Ensemble Modelling tool. We modelled incidence from officially reported cases, and adjusted our raw estimates for under-reporting based on published estimates of expansion factors. In total, we had 1780 country-years of mortality data from 130 countries, 1636 country-years of dengue case reports from 76 countries, and expansion factor estimates for 14 countries. FINDINGS We estimated an average of 9221 dengue deaths per year between 1990 and 2013, increasing from a low of 8277 (95% uncertainty estimate 5353-10 649) in 1992, to a peak of 11 302 (6790-13 722) in 2010. This yielded a total of 576 900 (330 000-701 200) years of life lost to premature mortality attributable to dengue in 2013. The incidence of dengue increased greatly between 1990 and 2013, with the number of cases more than doubling every decade, from 8·3 million (3·3 million-17·2 million) apparent cases in 1990, to 58·4 million (23·6 million-121·9 million) apparent cases in 2013. When accounting for disability from moderate and severe acute dengue, and post-dengue chronic fatigue, 566 000 (186 000-1 415 000) years lived with disability were attributable to dengue in 2013. Considering fatal and non-fatal outcomes together, dengue was responsible for 1·14 million (0·73 million-1·98 million) disability-adjusted life-years in 2013. INTERPRETATION Although lower than other estimates, our results offer more evidence that the true symptomatic incidence of dengue probably falls within the commonly cited range of 50 million to 100 million cases per year. Our mortality estimates are lower than those presented elsewhere and should be considered in light of the totality of evidence suggesting that dengue mortality might, in fact, be substantially higher. FUNDING Bill & Melinda Gates Foundation.',\n",
              "  0.003328621097218323),\n",
              " ('Native mammalian prions exist in self-propagating strains that exhibit distinctive clinical, pathological and biochemical characteristics. Prion strain diversity is associated with variations in PrP(Sc) conformation, but it remains unknown precisely which physical properties of the PrP(Sc) molecules are required to encipher mammalian prion strain phenotypes. In this study, we subjected prion-infected brain homogenates derived from three different hamster scrapie strains to either (i) proteinase K digestion or (ii) sonication, and inoculated the modified samples into normal hamsters. The results show that the strain-specific clinical features and neuropathological profiles of inoculated animals were not affected by either treatment. Similarly, the strain-dependent biochemical characteristics of the PrP(Sc) molecules (including electrophoretic mobility, glycoform composition, conformational stability and susceptibility to protease digestion) in infected animals were unaffected by either proteolysis or sonication of the original inocula. These results indicate that the infectious strain properties of native prions do not appear to be altered by PrP(Sc) disaggregation, and that maintenance of such properties does not require the N-domain (approximately residues 23-90) of the protease-resistant PrP(Sc) molecules or protease-sensitive PrP(Sc) molecules.',\n",
              "  0.002898264863585665),\n",
              " (\"The prion protein (PrP(C)) is highly expressed in the nervous system and critically involved in prion diseases where it misfolds into pathogenic PrP(Sc). Moreover, it has been suggested as a receptor mediating neurotoxicity in common neurodegenerative proteinopathies such as Alzheimer's disease. PrP(C) is shed at the plasma membrane by the metalloprotease ADAM10, yet the impact of this on prion disease remains enigmatic. Employing conditional knockout mice, we show that depletion of ADAM10 in forebrain neurons leads to posttranslational increase of PrP(C) levels. Upon prion infection of these mice, clinical, biochemical, and morphological data reveal that lack of ADAM10 significantly reduces incubation times and increases PrP(Sc) formation. In contrast, spatiotemporal analysis indicates that absence of shedding impairs spread of prion pathology. Our data support a dual role for ADAM10-mediated shedding and highlight the role of proteolytic processing in prion disease.\",\n",
              "  0.002415220719654721),\n",
              " ('BACKGROUND The World Health Organization (WHO) released revised guidelines in 2015 recommending that all people living with HIV, regardless of CD4 count, initiate antiretroviral therapy (ART) upon diagnosis. However, few studies have projected the global resources needed for rapid scale-up of ART. Under the Health Policy Project, we conducted modeling analyses for 97 countries to estimate eligibility for and numbers on ART from 2015 to 2020, along with the facility-level financial resources required. We compared the estimated financial requirements to estimated funding available. METHODS AND FINDINGS Current coverage levels and future need for treatment were based on country-specific epidemiological and demographic data. Simulated annual numbers of individuals on treatment were derived from three scenarios: (1) continuation of countries\\' current policies of eligibility for ART, (2) universal adoption of aspects of the WHO 2013 eligibility guidelines, and (3) expanded eligibility as per the WHO 2015 guidelines and meeting the Joint United Nations Programme on HIV/AIDS \"90-90-90\" ART targets. We modeled uncertainty in the annual resource requirements for antiretroviral drugs, laboratory tests, and facility-level personnel and overhead. We estimate that 25.7 (95% CI 25.5, 26.0) million adults and 1.57 (95% CI 1.55, 1.60) million children could receive ART by 2020 if countries maintain current eligibility plans and increase coverage based on historical rates, which may be ambitious. If countries uniformly adopt aspects of the WHO 2013 guidelines, 26.5 (95% CI 26.0 27.0) million adults and 1.53 (95% CI 1.52, 1.55) million children could be on ART by 2020. Under the 90-90-90 scenario, 30.4 (95% CI 30.1, 30.7) million adults and 1.68 (95% CI 1.63, 1.73) million children could receive treatment by 2020. The facility-level financial resources needed for scaling up ART in these countries from 2015 to 2020 are estimated to be US$45.8 (95% CI 45.4, 46.2) billion under the current scenario, US$48.7 (95% CI 47.8, 49.6) billion under the WHO 2013 scenario, and US$52.5 (95% CI 51.4, 53.6) billion under the 90-90-90 scenario. After projecting recent external and domestic funding trends, the estimated 6-y financing gap ranges from US$19.8 billion to US$25.0 billion, depending on the costing scenario and the U.S. President\\'s Emergency Plan for AIDS Relief contribution level, with the gap for ART commodities alone ranging from US$14.0 to US$16.8 billion. The study is limited by excluding above-facility and other costs essential to ART service delivery and by the availability and quality of country- and region-specific data. CONCLUSIONS The projected number of people receiving ART across three scenarios suggests that countries are unlikely to meet the 90-90-90 treatment target (81% of people living with HIV on ART by 2020) unless they adopt a test-and-offer approach and increase ART coverage. Our results suggest that future resource needs for ART scale-up are smaller than stated elsewhere but still significantly threaten the sustainability of the global HIV response without additional resource mobilization from domestic or innovative financing sources or efficiency gains. As the world moves towards adopting the WHO 2015 guidelines, advances in technology, including the introduction of lower-cost, highly effective antiretroviral regimens, whose value are assessed here, may prove to be \"game changers\" that allow more people to be on ART with the resources available.',\n",
              "  0.0018156115075736309),\n",
              " ('BACKGROUND Age-specific effects of mammographic screening, and the timing of such effects, are a matter of debate. The results of the UK Age trial, which compared the effect of invitation to annual mammographic screening from age 40 years with commencement of screening at age 50 years on breast cancer mortality, have been reported at 10 years of follow-up and showed no significant difference in mortality between the trial groups. Here, we report the results of the UK Age trial after 17 years of follow-up. METHODS Women aged 39-41 from 23 UK NHS Breast Screening Programme units years were randomly assigned by individual randomisation (1:2) to either an intervention group offered annual screening by mammography up to and including the calendar year of their 48th birthday or to a control group receiving usual medical care (invited for screening at age 50 years and every 3 years thereafter). Both groups were stratified by general practice. We compared breast cancer incidence and mortality by time since randomisation. Analyses included all women randomly assigned who could be traced with the National Health Service Central Register and who had not died or emigrated before entry. The primary outcome measures were mortality from breast cancer (defined as deaths with breast cancer coded as the underlying cause of death) and breast cancer incidence, including in-situ, invasive, and total incidence. Because there is an interest in the timing of the mortality effect, we analysed the results in different follow-up periods. This trial is registered, number ISRCTN24647151. FINDINGS Between Oct 14, 1990, and Sept 25, 1997, 160 921 participants were randomly assigned; 53 883 women in the intervention group and 106 953 assigned to usual medical care were included in this analysis. After a median follow-up of 17 years (IQR 16·8-18·8), the rate ratio (RR) for breast cancer mortality was 0·88 (95% CI 0·74-1·04) from tumours diagnosed during the intervention phase. A significant reduction in breast cancer mortality was noted in the intervention group compared with the control group in the first 10 years after diagnosis (RR 0·75, 0·58-0·97) but not thereafter (RR 1·02, 0·80-1·30) from tumours diagnosed during the intervention phase. The overall breast cancer incidence during 17 year follow-up was similar between the intervention group and the control group (RR 0·98, 0·93-1·04). INTERPRETATION Our results support an early reduction in mortality from breast cancer with annual mammography screening in women aged 40-49 years. Further data are needed to fully understand long-term effects. Cumulative incidence figures suggest at worst a small amount of overdiagnosis. FUNDING National Institute for Health Research Health Technology Assessment programme and the American Cancer Society. Past funding was received from the Medical Research Council, Cancer Research UK, the UK Department of Health, and the US National Cancer Institute.',\n",
              "  0.0016876296740242053),\n",
              " ('Prions are the protein-based infectious agents responsible for prion diseases. Environmental prion contamination has been implicated in disease transmission. Here, we analyzed the binding and retention of infectious prion protein (PrP(Sc)) to plants. Small quantities of PrP(Sc) contained in diluted brain homogenate or in excretory materials (urine and feces) can bind to wheat grass roots and leaves. Wild-type hamsters were efficiently infected by ingestion of prion-contaminated plants. The prion-plant interaction occurs with prions from diverse origins, including chronic wasting disease. Furthermore, leaves contaminated by spraying with a prion-containing preparation retained PrP(Sc) for several weeks in the living plant. Finally, plants can uptake prions from contaminated soil and transport them to aerial parts of the plant (stem and leaves). These findings demonstrate that plants can efficiently bind infectious prions and act as carriers of infectivity, suggesting a possible role of environmental prion contamination in the horizontal transmission of the disease.',\n",
              "  0.0014491324317928325),\n",
              " ('Infectious prions propagate from peripheral entry sites into the central nervous system (CNS), where they cause progressive neurodegeneration that ultimately leads to death. Yet the pathogenesis of prion disease can vary dramatically depending on the strain, or conformational variant of the aberrantly folded and aggregated protein, PrP(Sc). Although most prion strains invade the CNS, some prion strains cannot gain entry and do not cause clinical signs of disease. The conformational basis for this remarkable variation in the pathogenesis among strains is unclear. Using mouse-adapted prion strains, here we show that highly neuroinvasive prion strains primarily form diffuse aggregates in brain and are noncongophilic, conformationally unstable in denaturing conditions, and lead to rapidly lethal disease. These neuroinvasive strains efficiently generate PrP(Sc) over short incubation periods. In contrast, the weakly neuroinvasive prion strains form large fibrillary plaques and are stable, congophilic, and inefficiently generate PrP(Sc) over long incubation periods. Overall, these results indicate that the most neuroinvasive prion strains are also the least stable, and support the concept that the efficient replication and unstable nature of the most rapidly converting prions may be a feature linked to their efficient spread into the CNS.',\n",
              "  0.0014491324317928325),\n",
              " ('OBJECTIVE To assess the robustness of patient responses to a new national survey of patient experience as a basis for providing financial incentives to doctors. DESIGN Analysis of the representativeness of the respondents to the GP Patient Survey compared with those who were sampled (5.5 million patients registered with 8273 general practices in England in January 2009) and with the general population. Analysis of non-response bias looked at the relation between practice response rates and scores on the survey. Analysis of the reliability of the survey estimated the proportion of the variance of practice scores attributable to true differences between practices. RESULTS The overall response rate was 38.2% (2.2 million responses), which is comparable to that in surveys using similar methodology in the UK. Men, young adults, and people living in deprived areas were under-represented among respondents. However, for questions related to pay for performance, there was no systematic association between response rates and questionnaire scores. Two questions which triggered payments to general practitioners were reliable measures of practice performance, with average practice-level reliability coefficients of 93.2% and 95.0%. Less than 3% and 0.5% of practices had fewer than the number of responses required to achieve conventional reliability levels of 90% and 70%. A change to the payment formula in 2009 resulted in an increase in the average impact of random variation in patient scores on payments to general practitioners compared with payments made in 2007 and 2008. CONCLUSIONS There is little evidence to support the concern of some general practitioners that low response rates and selective non-response bias have led to systematic unfairness in payments attached to questionnaire scores. The study raises issues relating to the validity and reliability of payments based on patient surveys and provides lessons for the UK and for other countries considering the use of patient experience as part of pay for performance schemes.',\n",
              "  0.001280255705467559)]"
            ]
          },
          "metadata": {},
          "execution_count": 41
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ch29k9SLwFDB"
      },
      "source": [
        "## VSM combo test"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UnYiN_e9dXXc",
        "outputId": "ee295adc-cd94-4aa4-f3a0-0a7760b4669d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['1 in 5 million in UK have abnormal PrP positivity.']\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([0, 0, 0, ..., 0, 0, 0])"
            ]
          },
          "metadata": {},
          "execution_count": 45
        }
      ],
      "source": [
        "## VSM retrieval\n",
        "print([queries[query_num][1]])\n",
        "q_vector"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 53,
      "metadata": {
        "id": "F7F3a2EewwcS",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "683bc43d-1d31-4a3a-e141-682ad72568ad"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[('OBJECTIVES To carry out a further survey of archived appendix samples to understand better the differences between existing estimates of the prevalence of subclinical infection with prions after the bovine spongiform encephalopathy epizootic and to see whether a broader birth cohort was affected, and to understand better the implications for the management of blood and blood products and for the handling of surgical instruments. DESIGN Irreversibly unlinked and anonymised large scale survey of archived appendix samples. SETTING Archived appendix samples from the pathology departments of 41 UK hospitals participating in the earlier survey, and additional hospitals in regions with lower levels of participation in that survey. SAMPLE 32,441 archived appendix samples fixed in formalin and embedded in paraffin and tested for the presence of abnormal prion protein (PrP). RESULTS Of the 32,441 appendix samples 16 were positive for abnormal PrP, indicating an overall prevalence of 493 per million population (95% confidence interval 282 to 801 per million). The prevalence in those born in 1941-60 (733 per million, 269 to 1596 per million) did not differ significantly from those born between 1961 and 1985 (412 per million, 198 to 758 per million) and was similar in both sexes and across the three broad geographical areas sampled. Genetic testing of the positive specimens for the genotype at PRNP codon 129 revealed a high proportion that were valine homozygous compared with the frequency in the normal population, and in stark contrast with confirmed clinical cases of vCJD, all of which were methionine homozygous at PRNP codon 129. CONCLUSIONS This study corroborates previous studies and suggests a high prevalence of infection with abnormal PrP, indicating vCJD carrier status in the population compared with the 177 vCJD cases to date. These findings have important implications for the management of blood and blood products and for the handling of surgical instruments.',\n",
              "  0.0034240379091171245),\n",
              " ('We report a case of preclinical variant Creutzfeldt-Jakob disease (vCJD) in a patient who died from a non-neurological disorder 5 years after receiving a blood transfusion from a donor who subsequently developed vCJD. Protease-resistant prion protein (PrP(res)) was detected by western blot, paraffin-embedded tissue blot, and immunohistochemistry in the spleen, but not in the brain. Immunohistochemistry for prion protein was also positive in a cervical lymph node. The patient was a heterozygote at codon 129 of PRNP, suggesting that susceptibility to vCJD infection is not confined to the methionine homozygous PRNP genotype. These findings have major implications for future estimates and surveillance of vCJD in the UK.',\n",
              "  0.0018811298258289446),\n",
              " (\"The prion protein (PrP(C)) is highly expressed in the nervous system and critically involved in prion diseases where it misfolds into pathogenic PrP(Sc). Moreover, it has been suggested as a receptor mediating neurotoxicity in common neurodegenerative proteinopathies such as Alzheimer's disease. PrP(C) is shed at the plasma membrane by the metalloprotease ADAM10, yet the impact of this on prion disease remains enigmatic. Employing conditional knockout mice, we show that depletion of ADAM10 in forebrain neurons leads to posttranslational increase of PrP(C) levels. Upon prion infection of these mice, clinical, biochemical, and morphological data reveal that lack of ADAM10 significantly reduces incubation times and increases PrP(Sc) formation. In contrast, spatiotemporal analysis indicates that absence of shedding impairs spread of prion pathology. Our data support a dual role for ADAM10-mediated shedding and highlight the role of proteolytic processing in prion disease.\",\n",
              "  0.00162611725812254),\n",
              " ('Native mammalian prions exist in self-propagating strains that exhibit distinctive clinical, pathological and biochemical characteristics. Prion strain diversity is associated with variations in PrP(Sc) conformation, but it remains unknown precisely which physical properties of the PrP(Sc) molecules are required to encipher mammalian prion strain phenotypes. In this study, we subjected prion-infected brain homogenates derived from three different hamster scrapie strains to either (i) proteinase K digestion or (ii) sonication, and inoculated the modified samples into normal hamsters. The results show that the strain-specific clinical features and neuropathological profiles of inoculated animals were not affected by either treatment. Similarly, the strain-dependent biochemical characteristics of the PrP(Sc) molecules (including electrophoretic mobility, glycoform composition, conformational stability and susceptibility to protease digestion) in infected animals were unaffected by either proteolysis or sonication of the original inocula. These results indicate that the infectious strain properties of native prions do not appear to be altered by PrP(Sc) disaggregation, and that maintenance of such properties does not require the N-domain (approximately residues 23-90) of the protease-resistant PrP(Sc) molecules or protease-sensitive PrP(Sc) molecules.',\n",
              "  0.001609261997218953),\n",
              " ('Prions are the protein-based infectious agents responsible for prion diseases. Environmental prion contamination has been implicated in disease transmission. Here, we analyzed the binding and retention of infectious prion protein (PrP(Sc)) to plants. Small quantities of PrP(Sc) contained in diluted brain homogenate or in excretory materials (urine and feces) can bind to wheat grass roots and leaves. Wild-type hamsters were efficiently infected by ingestion of prion-contaminated plants. The prion-plant interaction occurs with prions from diverse origins, including chronic wasting disease. Furthermore, leaves contaminated by spraying with a prion-containing preparation retained PrP(Sc) for several weeks in the living plant. Finally, plants can uptake prions from contaminated soil and transport them to aerial parts of the plant (stem and leaves). These findings demonstrate that plants can efficiently bind infectious prions and act as carriers of infectivity, suggesting a possible role of environmental prion contamination in the horizontal transmission of the disease.',\n",
              "  0.001432821723301387),\n",
              " ('OBJECTIVE To assess the robustness of patient responses to a new national survey of patient experience as a basis for providing financial incentives to doctors. DESIGN Analysis of the representativeness of the respondents to the GP Patient Survey compared with those who were sampled (5.5 million patients registered with 8273 general practices in England in January 2009) and with the general population. Analysis of non-response bias looked at the relation between practice response rates and scores on the survey. Analysis of the reliability of the survey estimated the proportion of the variance of practice scores attributable to true differences between practices. RESULTS The overall response rate was 38.2% (2.2 million responses), which is comparable to that in surveys using similar methodology in the UK. Men, young adults, and people living in deprived areas were under-represented among respondents. However, for questions related to pay for performance, there was no systematic association between response rates and questionnaire scores. Two questions which triggered payments to general practitioners were reliable measures of practice performance, with average practice-level reliability coefficients of 93.2% and 95.0%. Less than 3% and 0.5% of practices had fewer than the number of responses required to achieve conventional reliability levels of 90% and 70%. A change to the payment formula in 2009 resulted in an increase in the average impact of random variation in patient scores on payments to general practitioners compared with payments made in 2007 and 2008. CONCLUSIONS There is little evidence to support the concern of some general practitioners that low response rates and selective non-response bias have led to systematic unfairness in payments attached to questionnaire scores. The study raises issues relating to the validity and reliability of payments based on patient surveys and provides lessons for the UK and for other countries considering the use of patient experience as part of pay for performance schemes.',\n",
              "  0.0014252983461446479),\n",
              " ('Infectious prions propagate from peripheral entry sites into the central nervous system (CNS), where they cause progressive neurodegeneration that ultimately leads to death. Yet the pathogenesis of prion disease can vary dramatically depending on the strain, or conformational variant of the aberrantly folded and aggregated protein, PrP(Sc). Although most prion strains invade the CNS, some prion strains cannot gain entry and do not cause clinical signs of disease. The conformational basis for this remarkable variation in the pathogenesis among strains is unclear. Using mouse-adapted prion strains, here we show that highly neuroinvasive prion strains primarily form diffuse aggregates in brain and are noncongophilic, conformationally unstable in denaturing conditions, and lead to rapidly lethal disease. These neuroinvasive strains efficiently generate PrP(Sc) over short incubation periods. In contrast, the weakly neuroinvasive prion strains form large fibrillary plaques and are stable, congophilic, and inefficiently generate PrP(Sc) over long incubation periods. Overall, these results indicate that the most neuroinvasive prion strains are also the least stable, and support the concept that the efficient replication and unstable nature of the most rapidly converting prions may be a feature linked to their efficient spread into the CNS.',\n",
              "  0.0013915304062191351),\n",
              " ('Hemolysates of erythrocytes from more than a quarter million people in Alabama were electrophoresed on cellulose acetate, pH 8.4, and those samples exhibiting an abnormality were also electrophoresed in citrate agar, pH 6.0. The globin chains of mutants other than Hb S and C were electrophoresed in urea-mercaptoethanol buffers at both pH 8.9 and pH 6.0, and 60 of them were also analyzed structurally. Of about 6000 samples from whites, only three contained abnormal hemoglobins--Hb D Los Angeles, Hb J Baltimore, and one unidentified. Of 249,000 samples from blacks, about 29,000 contained electrophoretically detectable abnormalities, most of them associated with Hb S or C, present in a frequency of about 9% and 3%, respectively. About 1000 samples resolved into patterns of potential clinical significance. Twenty other mutant hemoglobins were detected, in various genetic combinations in 164 kindreds; four of these-Hb Alabama, Montgomery, Titusville, and Mobile--were previously unknown. The methods used are rapid, economical, and well suited for large scale surveys. They provide highly specific characterizations of many mutant hemoglobins, and no discrepancies were found between the presumptive identifications based on these characterizations and the definitive identifications obtained from structural analyses.',\n",
              "  0.001146764201047363),\n",
              " ('A considerable subgroup of patients with early breast cancer does not address benefits of anthracycline based chemotherapy. The aim of this retrospective study was to investigate the effect of microvessel density (MVD) and status of p53 protein on 5-year disease free survival (DFS) in the group of breast cancer patients treated with anthracyclines in adjuvant setting. Correlations between MVD, p53 status and other clinicopathological parameters were also assessed. MVD and p53 status were analyzed immunohistochemically in the group of 172 women with breast cancer in clinical stage T1-2, N1-N2, M0. There were 123 tumors (71.5 %) with lower MVD (≤ 214.8 microvesells/mm(2)) and 49 (28.5 %) with higher MVD (>214.8 microvesells/mm(2)). The proportion of higher MVD tumors significantly increased in N2 (P = 0.000) and in estrogen (P = 0.046) or progesterone receptors (P = 0.029) negative tumors. p53 positivity was indicated in 50 cancers (29.1 %) and was significantly associated with higher grade (P = 0.000), steroid receptors negativity (P = 0.000), cytokeratin5/6 positivity (P = 0.026), topoisomerase IIα overexpression (P = 0.005) and higher proliferation rate (P = 0.001). In univariate analysis, higher MVD (P = 0.016) and p53 negativity (P = 0.023) were significantly related with longer DFS (median follow-up 36 months). In multivariate Cox regression analysis MVD was independently associated with DFS. These data suggest that higher MVD is favourable prognostic factors for early advanced breast cancer patients after adjuvant anthracycline based chemotherapy.',\n",
              "  0.0010817303365919912),\n",
              " ('OBJECTIVE To determine whether myelin oligodendrocyte glycoprotein antibodies (MOG-Abs) were predictive of a demyelination phenotype in children presenting with acquired demyelinating syndrome (ADS). METHOD Sixty-five children with a first episode of ADS (12 acute disseminated encephalomyelitis, 24 optic neuritis, 18 transverse myelitis, 11 other clinically isolated syndrome) were identified from 2 national demyelination programs in the United Kingdom and France. Acute serum samples were tested for MOG-Abs by cell-based assay. Antibodies were used to predict diagnosis of multiple sclerosis (MS) at 1 year. RESULTS Twenty-three of 65 (35%) children had MOG-Abs. Antibody-positive and antibody-negative patients were not clinically different at presentation, but identification of MOG-Abs predicted a non-MS course at 1-year follow-up: only 2/23 (9%) MOG-Ab-positive patients were diagnosed with MS compared to 16/42 (38%) MOG-Ab-negative patients (p = 0.019, Fisher exact test). Antibody positivity at outset was a useful predictor for a non-MS disease course, with a positive predictive value of 91% (95% confidence interval [CI] 72-99), negative predictive value of 38% (95% CI 24-54), positive likelihood ratio of 4.02 (CI 1.0-15.4), and odds ratio of 6.5 (CI 1.3-31.3). CONCLUSIONS MOG-Abs are found at presentation in 35% of patients with childhood ADS, across a range of demyelinating disorders. Antibody positivity can be useful in predicting a non-MS disease course at onset.',\n",
              "  0.0010746583933246739)]"
            ]
          },
          "metadata": {},
          "execution_count": 53
        }
      ],
      "source": [
        "doc_vector = np.array(bm25_idf)\n",
        "\n",
        "BM25_VSM_scores = np.dot(doc_vector, q_vector)/(norm(doc_vector)*norm(q_vector))\n",
        "sorted(zip(documents,BM25_VSM_scores), key = lambda tup:tup[1], reverse=True)[:10]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uKRnSigB5ZHp"
      },
      "source": [
        "https://ai.googleblog.com/2018/05/advances-in-semantic-textual-similarity.html\n",
        "https://www.tensorflow.org/hub/tutorials/semantic_similarity_with_tf_hub_universal_encoder "
      ]
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [
        "8CvP86nnwTuw",
        "7OvjLQnH80JX",
        "Ch29k9SLwFDB"
      ],
      "name": "Search_engine_v0.0.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyOaOMDdFNZN9UakZgMQiybT",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}