{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Search_engine_v0.0.ipynb","provenance":[{"file_id":"1PgPZf47UxhUdOgrcPmZZM99Q58IcLg1k","timestamp":1647958891868}],"collapsed_sections":["MhN0gWKiBGiQ"],"toc_visible":true,"authorship_tag":"ABX9TyP45C5+pJ3qwvqE9Lr9O36x"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Data loading testing\n","see https://ir-datasets.com/beir.html#beir/scifact  \n","and https://ir-datasets.com/python.html"],"metadata":{"id":"XeS4WON90flx"}},{"cell_type":"code","execution_count":1,"metadata":{"id":"5Qh8yMNt6QeI","executionInfo":{"status":"ok","timestamp":1647985248465,"user_tz":0,"elapsed":20275,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}}},"outputs":[],"source":["%%capture\n","!pip install --upgrade ir_datasets\n","\n","import pandas as pd\n","from sklearn.feature_extraction.text import CountVectorizer\n","import numpy as np\n","import ir_datasets"]},{"cell_type":"markdown","source":["## Scifact test"],"metadata":{"id":"1YxFIFKTBNLM"}},{"cell_type":"code","source":["dataset = ir_datasets.load(\"beir/scifact/train\")\n","\n","print(dataset.docs_count(), 'documents with:', dataset.docs_cls().__annotations__)\n","print(dataset.queries_count(), 'queries with:', dataset.queries_cls().__annotations__)\n","print(dataset.qrels_count(), 'qrels with:',dataset.qrels_cls().__annotations__)"],"metadata":{"id":"i1g2Gf2LjToC","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1647985251950,"user_tz":0,"elapsed":5,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}},"outputId":"226b3659-26ea-4066-fc92-9b0e43a493d4"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["5183 documents with: OrderedDict([('doc_id', <class 'str'>), ('text', <class 'str'>), ('title', <class 'str'>)])\n","809 queries with: OrderedDict([('query_id', <class 'str'>), ('text', <class 'str'>)])\n","919 qrels with: OrderedDict([('query_id', <class 'str'>), ('doc_id', <class 'str'>), ('relevance', <class 'int'>), ('iteration', <class 'str'>)])\n"]}]},{"cell_type":"code","source":["%%capture\n","#this actually downloads the data, muted to avoid clutter\n","for doc in dataset.docs_iter()[:1]: \n","    None\n","for q in dataset.queries_iter():\n","    None\n","for qrel in dataset.qrels_iter():\n","    None"],"metadata":{"id":"WIDi1JwIjh7R","executionInfo":{"status":"ok","timestamp":1647985398363,"user_tz":0,"elapsed":212,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}}},"execution_count":11,"outputs":[]},{"cell_type":"code","source":["corpus=[]\n","for doc in dataset.docs_iter():\n","    corpus.append([doc[0],doc[1],doc[2]])\n","\n","documents = [doc[1] for doc in corpus] # only taking the text for simplicity, but we'll need the doc_id for the evaluation"],"metadata":{"id":"pkhFTCerkCNh","executionInfo":{"status":"ok","timestamp":1647985400960,"user_tz":0,"elapsed":227,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}}},"execution_count":12,"outputs":[]},{"cell_type":"code","source":["vectorizer = CountVectorizer(stop_words='english', strip_accents='ascii')\n","documents_vectorized = vectorizer.fit_transform(documents)\n","vocabulary = vectorizer.get_feature_names_out()\n","dataframe = pd.DataFrame(documents_vectorized.toarray(), columns=vocabulary)\n","dataframe"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":424},"id":"G3E7sQ5_kg9x","executionInfo":{"status":"ok","timestamp":1647985409255,"user_tz":0,"elapsed":2180,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}},"outputId":"1ec1e09d-14f1-4eff-bced-5103219a908b"},"execution_count":13,"outputs":[{"output_type":"execute_result","data":{"text/plain":["      00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n","0      0    0     0      0      0       0      0      0      0      0  ...   \n","1      0    0     0      0      0       0      0      0      0      0  ...   \n","2      0    0     0      0      0       0      0      0      0      0  ...   \n","3      0    0     0      0      0       0      0      0      0      0  ...   \n","4      0    0     0      0      0       0      0      0      0      0  ...   \n","...   ..  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n","5178   0    0     0      0      0       0      0      0      0      0  ...   \n","5179   0    0     0      0      0       0      0      0      0      0  ...   \n","5180   0    0     0      0      0       0      0      0      0      0  ...   \n","5181   0    0     0      0      0       0      0      0      0      0  ...   \n","5182   0    0     0      0      0       0      0      0      0      0  ...   \n","\n","      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin  zz  zzw  \\\n","0            0       0        0        0           0        0      0   0    0   \n","1            0       0        0        0           0        0      0   0    0   \n","2            0       0        0        0           0        0      0   0    0   \n","3            0       0        0        0           0        0      0   0    0   \n","4            0       0        0        0           0        0      0   0    0   \n","...        ...     ...      ...      ...         ...      ...    ...  ..  ...   \n","5178         0       0        0        0           0        0      0   0    0   \n","5179         0       0        0        0           0        0      0   0    0   \n","5180         0       0        0        0           0        0      0   0    0   \n","5181         0       0        0        0           0        0      0   0    0   \n","5182         0       0        0        0           0        0      0   0    0   \n","\n","      zzz  \n","0       0  \n","1       0  \n","2       0  \n","3       0  \n","4       0  \n","...   ...  \n","5178    0  \n","5179    0  \n","5180    0  \n","5181    0  \n","5182    0  \n","\n","[5183 rows x 35568 columns]"],"text/html":["\n","  <div id=\"df-f7ff6b26-e46c-42d2-a972-e9f6a4303858\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>00</th>\n","      <th>000</th>\n","      <th>0000</th>\n","      <th>00001</th>\n","      <th>00002</th>\n","      <th>000035</th>\n","      <th>00004</th>\n","      <th>00005</th>\n","      <th>00006</th>\n","      <th>00008</th>\n","      <th>...</th>\n","      <th>zygosity</th>\n","      <th>zygote</th>\n","      <th>zygotes</th>\n","      <th>zygotic</th>\n","      <th>zymography</th>\n","      <th>zymosan</th>\n","      <th>zyxin</th>\n","      <th>zz</th>\n","      <th>zzw</th>\n","      <th>zzz</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>5178</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>5179</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>5180</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>5181</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","    <tr>\n","      <th>5182</th>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>...</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","      <td>0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5183 rows × 35568 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f7ff6b26-e46c-42d2-a972-e9f6a4303858')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-f7ff6b26-e46c-42d2-a972-e9f6a4303858 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-f7ff6b26-e46c-42d2-a972-e9f6a4303858');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":13}]},{"cell_type":"code","source":["# Defining all the variables we need:\n","k_1 = 1.2\n","b = 0.8\n","dfs = (dataframe > 0).sum(axis=0) # doc frequency\n","N = dataframe.shape[0] # total number of docs\n","idfs = np.log10(N/dfs) # inverse doc frequency\n","#dls = [len(d.split(' ')) for d in documents] # considering all words in doc\n","dls = dataframe.sum(axis=1).tolist() # considering words minus stop words in doc (better option)\n","avgdl = np.mean(dls) # single value, mean doc length (minus stop words)\n","\n","# Applying the BM25 formula:\n","numerator = np.array((k_1 + 1) * dataframe)\n","denominator = np.array(k_1 *((1 - b) + b * (dls / avgdl))).reshape(N,1) + np.array(dataframe)\n","BM25_tf = numerator / denominator\n","idfs = np.array(idfs)\n","BM25_score = BM25_tf * idfs\n","\n","bm25_idf = pd.DataFrame(BM25_score, columns=vocabulary)\n","bm25_idf"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":424},"id":"OtpbYAsEk7up","executionInfo":{"status":"ok","timestamp":1647985416823,"user_tz":0,"elapsed":5306,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}},"outputId":"104e6332-4c18-4387-aad0-ce546b52e39f"},"execution_count":14,"outputs":[{"output_type":"execute_result","data":{"text/plain":["       00  000  0000  00001  00002  000035  00004  00005  00006  00008  ...  \\\n","0     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","1     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","2     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","3     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","4     0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","...   ...  ...   ...    ...    ...     ...    ...    ...    ...    ...  ...   \n","5178  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","5179  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","5180  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","5181  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","5182  0.0  0.0   0.0    0.0    0.0     0.0    0.0    0.0    0.0    0.0  ...   \n","\n","      zygosity  zygote  zygotes  zygotic  zymography  zymosan  zyxin   zz  \\\n","0          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","1          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","2          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","3          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","4          0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","...        ...     ...      ...      ...         ...      ...    ...  ...   \n","5178       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","5179       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","5180       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","5181       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","5182       0.0     0.0      0.0      0.0         0.0      0.0    0.0  0.0   \n","\n","      zzw  zzz  \n","0     0.0  0.0  \n","1     0.0  0.0  \n","2     0.0  0.0  \n","3     0.0  0.0  \n","4     0.0  0.0  \n","...   ...  ...  \n","5178  0.0  0.0  \n","5179  0.0  0.0  \n","5180  0.0  0.0  \n","5181  0.0  0.0  \n","5182  0.0  0.0  \n","\n","[5183 rows x 35568 columns]"],"text/html":["\n","  <div id=\"df-3992851c-b83a-4029-a897-44dd1742e555\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>00</th>\n","      <th>000</th>\n","      <th>0000</th>\n","      <th>00001</th>\n","      <th>00002</th>\n","      <th>000035</th>\n","      <th>00004</th>\n","      <th>00005</th>\n","      <th>00006</th>\n","      <th>00008</th>\n","      <th>...</th>\n","      <th>zygosity</th>\n","      <th>zygote</th>\n","      <th>zygotes</th>\n","      <th>zygotic</th>\n","      <th>zymography</th>\n","      <th>zymosan</th>\n","      <th>zyxin</th>\n","      <th>zz</th>\n","      <th>zzw</th>\n","      <th>zzz</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>5178</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>5179</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>5180</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>5181</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","    <tr>\n","      <th>5182</th>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>...</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","      <td>0.0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5183 rows × 35568 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3992851c-b83a-4029-a897-44dd1742e555')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-3992851c-b83a-4029-a897-44dd1742e555 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-3992851c-b83a-4029-a897-44dd1742e555');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":14}]},{"cell_type":"code","source":["query_num = 1 #query number to test\n","queries = []\n","for q in dataset.queries_iter(): # reading in all the queries\n","    queries.append([q[0],q[1]])\n","print([queries[query_num][1]])\n","\n","vectorizer_q = CountVectorizer(stop_words='english', strip_accents='ascii')\n","query_vectorized = vectorizer_q.fit_transform([queries[query_num][1]])\n","q_terms = vectorizer_q.get_feature_names_out()\n","q_terms"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PotSpiIxltBy","executionInfo":{"status":"ok","timestamp":1647985420059,"user_tz":0,"elapsed":241,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}},"outputId":"6c94c3b6-57c2-4646-ca42-262b4735a35d"},"execution_count":15,"outputs":[{"output_type":"stream","name":"stdout","text":["['1 in 5 million in UK have abnormal PrP positivity.']\n"]},{"output_type":"execute_result","data":{"text/plain":["array(['abnormal', 'million', 'positivity', 'prp', 'uk'], dtype=object)"]},"metadata":{},"execution_count":15}]},{"cell_type":"code","source":["q_terms = [term for term in q_terms if term in bm25_idf.columns]\n","q_terms_only_df = bm25_idf[q_terms]\n","score_q_d = q_terms_only_df.sum(axis=1)\n","documents = [doc[1] for doc in corpus]\n","sorted(zip(documents,score_q_d.values), key = lambda tup:tup[1], reverse=True)[:10]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ypvq9CGBl6Is","executionInfo":{"status":"ok","timestamp":1647985425046,"user_tz":0,"elapsed":288,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}},"outputId":"9dc14577-ddd4-4222-98b4-82d3b216523e"},"execution_count":16,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[('OBJECTIVES To carry out a further survey of archived appendix samples to understand better the differences between existing estimates of the prevalence of subclinical infection with prions after the bovine spongiform encephalopathy epizootic and to see whether a broader birth cohort was affected, and to understand better the implications for the management of blood and blood products and for the handling of surgical instruments. DESIGN Irreversibly unlinked and anonymised large scale survey of archived appendix samples. SETTING Archived appendix samples from the pathology departments of 41 UK hospitals participating in the earlier survey, and additional hospitals in regions with lower levels of participation in that survey. SAMPLE 32,441 archived appendix samples fixed in formalin and embedded in paraffin and tested for the presence of abnormal prion protein (PrP). RESULTS Of the 32,441 appendix samples 16 were positive for abnormal PrP, indicating an overall prevalence of 493 per million population (95% confidence interval 282 to 801 per million). The prevalence in those born in 1941-60 (733 per million, 269 to 1596 per million) did not differ significantly from those born between 1961 and 1985 (412 per million, 198 to 758 per million) and was similar in both sexes and across the three broad geographical areas sampled. Genetic testing of the positive specimens for the genotype at PRNP codon 129 revealed a high proportion that were valine homozygous compared with the frequency in the normal population, and in stark contrast with confirmed clinical cases of vCJD, all of which were methionine homozygous at PRNP codon 129. CONCLUSIONS This study corroborates previous studies and suggests a high prevalence of infection with abnormal PrP, indicating vCJD carrier status in the population compared with the 177 vCJD cases to date. These findings have important implications for the management of blood and blood products and for the handling of surgical instruments.',\n","  11.537320382799532),\n"," ('We report a case of preclinical variant Creutzfeldt-Jakob disease (vCJD) in a patient who died from a non-neurological disorder 5 years after receiving a blood transfusion from a donor who subsequently developed vCJD. Protease-resistant prion protein (PrP(res)) was detected by western blot, paraffin-embedded tissue blot, and immunohistochemistry in the spleen, but not in the brain. Immunohistochemistry for prion protein was also positive in a cervical lymph node. The patient was a heterozygote at codon 129 of PRNP, suggesting that susceptibility to vCJD infection is not confined to the methionine homozygous PRNP genotype. These findings have major implications for future estimates and surveillance of vCJD in the UK.',\n","  6.3384804894944935),\n"," (\"The prion protein (PrP(C)) is highly expressed in the nervous system and critically involved in prion diseases where it misfolds into pathogenic PrP(Sc). Moreover, it has been suggested as a receptor mediating neurotoxicity in common neurodegenerative proteinopathies such as Alzheimer's disease. PrP(C) is shed at the plasma membrane by the metalloprotease ADAM10, yet the impact of this on prion disease remains enigmatic. Employing conditional knockout mice, we show that depletion of ADAM10 in forebrain neurons leads to posttranslational increase of PrP(C) levels. Upon prion infection of these mice, clinical, biochemical, and morphological data reveal that lack of ADAM10 significantly reduces incubation times and increases PrP(Sc) formation. In contrast, spatiotemporal analysis indicates that absence of shedding impairs spread of prion pathology. Our data support a dual role for ADAM10-mediated shedding and highlight the role of proteolytic processing in prion disease.\",\n","  5.479213806892906),\n"," ('Native mammalian prions exist in self-propagating strains that exhibit distinctive clinical, pathological and biochemical characteristics. Prion strain diversity is associated with variations in PrP(Sc) conformation, but it remains unknown precisely which physical properties of the PrP(Sc) molecules are required to encipher mammalian prion strain phenotypes. In this study, we subjected prion-infected brain homogenates derived from three different hamster scrapie strains to either (i) proteinase K digestion or (ii) sonication, and inoculated the modified samples into normal hamsters. The results show that the strain-specific clinical features and neuropathological profiles of inoculated animals were not affected by either treatment. Similarly, the strain-dependent biochemical characteristics of the PrP(Sc) molecules (including electrophoretic mobility, glycoform composition, conformational stability and susceptibility to protease digestion) in infected animals were unaffected by either proteolysis or sonication of the original inocula. These results indicate that the infectious strain properties of native prions do not appear to be altered by PrP(Sc) disaggregation, and that maintenance of such properties does not require the N-domain (approximately residues 23-90) of the protease-resistant PrP(Sc) molecules or protease-sensitive PrP(Sc) molecules.',\n","  5.42241988394522),\n"," ('Prions are the protein-based infectious agents responsible for prion diseases. Environmental prion contamination has been implicated in disease transmission. Here, we analyzed the binding and retention of infectious prion protein (PrP(Sc)) to plants. Small quantities of PrP(Sc) contained in diluted brain homogenate or in excretory materials (urine and feces) can bind to wheat grass roots and leaves. Wild-type hamsters were efficiently infected by ingestion of prion-contaminated plants. The prion-plant interaction occurs with prions from diverse origins, including chronic wasting disease. Furthermore, leaves contaminated by spraying with a prion-containing preparation retained PrP(Sc) for several weeks in the living plant. Finally, plants can uptake prions from contaminated soil and transport them to aerial parts of the plant (stem and leaves). These findings demonstrate that plants can efficiently bind infectious prions and act as carriers of infectivity, suggesting a possible role of environmental prion contamination in the horizontal transmission of the disease.',\n","  4.827903110869903),\n"," ('OBJECTIVE To assess the robustness of patient responses to a new national survey of patient experience as a basis for providing financial incentives to doctors. DESIGN Analysis of the representativeness of the respondents to the GP Patient Survey compared with those who were sampled (5.5 million patients registered with 8273 general practices in England in January 2009) and with the general population. Analysis of non-response bias looked at the relation between practice response rates and scores on the survey. Analysis of the reliability of the survey estimated the proportion of the variance of practice scores attributable to true differences between practices. RESULTS The overall response rate was 38.2% (2.2 million responses), which is comparable to that in surveys using similar methodology in the UK. Men, young adults, and people living in deprived areas were under-represented among respondents. However, for questions related to pay for performance, there was no systematic association between response rates and questionnaire scores. Two questions which triggered payments to general practitioners were reliable measures of practice performance, with average practice-level reliability coefficients of 93.2% and 95.0%. Less than 3% and 0.5% of practices had fewer than the number of responses required to achieve conventional reliability levels of 90% and 70%. A change to the payment formula in 2009 resulted in an increase in the average impact of random variation in patient scores on payments to general practitioners compared with payments made in 2007 and 2008. CONCLUSIONS There is little evidence to support the concern of some general practitioners that low response rates and selective non-response bias have led to systematic unfairness in payments attached to questionnaire scores. The study raises issues relating to the validity and reliability of payments based on patient surveys and provides lessons for the UK and for other countries considering the use of patient experience as part of pay for performance schemes.',\n","  4.802553037383037),\n"," ('Infectious prions propagate from peripheral entry sites into the central nervous system (CNS), where they cause progressive neurodegeneration that ultimately leads to death. Yet the pathogenesis of prion disease can vary dramatically depending on the strain, or conformational variant of the aberrantly folded and aggregated protein, PrP(Sc). Although most prion strains invade the CNS, some prion strains cannot gain entry and do not cause clinical signs of disease. The conformational basis for this remarkable variation in the pathogenesis among strains is unclear. Using mouse-adapted prion strains, here we show that highly neuroinvasive prion strains primarily form diffuse aggregates in brain and are noncongophilic, conformationally unstable in denaturing conditions, and lead to rapidly lethal disease. These neuroinvasive strains efficiently generate PrP(Sc) over short incubation periods. In contrast, the weakly neuroinvasive prion strains form large fibrillary plaques and are stable, congophilic, and inefficiently generate PrP(Sc) over long incubation periods. Overall, these results indicate that the most neuroinvasive prion strains are also the least stable, and support the concept that the efficient replication and unstable nature of the most rapidly converting prions may be a feature linked to their efficient spread into the CNS.',\n","  4.688771720724594),\n"," ('Hemolysates of erythrocytes from more than a quarter million people in Alabama were electrophoresed on cellulose acetate, pH 8.4, and those samples exhibiting an abnormality were also electrophoresed in citrate agar, pH 6.0. The globin chains of mutants other than Hb S and C were electrophoresed in urea-mercaptoethanol buffers at both pH 8.9 and pH 6.0, and 60 of them were also analyzed structurally. Of about 6000 samples from whites, only three contained abnormal hemoglobins--Hb D Los Angeles, Hb J Baltimore, and one unidentified. Of 249,000 samples from blacks, about 29,000 contained electrophoretically detectable abnormalities, most of them associated with Hb S or C, present in a frequency of about 9% and 3%, respectively. About 1000 samples resolved into patterns of potential clinical significance. Twenty other mutant hemoglobins were detected, in various genetic combinations in 164 kindreds; four of these-Hb Alabama, Montgomery, Titusville, and Mobile--were previously unknown. The methods used are rapid, economical, and well suited for large scale surveys. They provide highly specific characterizations of many mutant hemoglobins, and no discrepancies were found between the presumptive identifications based on these characterizations and the definitive identifications obtained from structural analyses.',\n","  3.8640302304421676),\n"," ('A considerable subgroup of patients with early breast cancer does not address benefits of anthracycline based chemotherapy. The aim of this retrospective study was to investigate the effect of microvessel density (MVD) and status of p53 protein on 5-year disease free survival (DFS) in the group of breast cancer patients treated with anthracyclines in adjuvant setting. Correlations between MVD, p53 status and other clinicopathological parameters were also assessed. MVD and p53 status were analyzed immunohistochemically in the group of 172 women with breast cancer in clinical stage T1-2, N1-N2, M0. There were 123 tumors (71.5 %) with lower MVD (≤ 214.8 microvesells/mm(2)) and 49 (28.5 %) with higher MVD (>214.8 microvesells/mm(2)). The proportion of higher MVD tumors significantly increased in N2 (P = 0.000) and in estrogen (P = 0.046) or progesterone receptors (P = 0.029) negative tumors. p53 positivity was indicated in 50 cancers (29.1 %) and was significantly associated with higher grade (P = 0.000), steroid receptors negativity (P = 0.000), cytokeratin5/6 positivity (P = 0.026), topoisomerase IIα overexpression (P = 0.005) and higher proliferation rate (P = 0.001). In univariate analysis, higher MVD (P = 0.016) and p53 negativity (P = 0.023) were significantly related with longer DFS (median follow-up 36 months). In multivariate Cox regression analysis MVD was independently associated with DFS. These data suggest that higher MVD is favourable prognostic factors for early advanced breast cancer patients after adjuvant anthracycline based chemotherapy.',\n","  3.6448981560117626),\n"," ('OBJECTIVE To determine whether myelin oligodendrocyte glycoprotein antibodies (MOG-Abs) were predictive of a demyelination phenotype in children presenting with acquired demyelinating syndrome (ADS). METHOD Sixty-five children with a first episode of ADS (12 acute disseminated encephalomyelitis, 24 optic neuritis, 18 transverse myelitis, 11 other clinically isolated syndrome) were identified from 2 national demyelination programs in the United Kingdom and France. Acute serum samples were tested for MOG-Abs by cell-based assay. Antibodies were used to predict diagnosis of multiple sclerosis (MS) at 1 year. RESULTS Twenty-three of 65 (35%) children had MOG-Abs. Antibody-positive and antibody-negative patients were not clinically different at presentation, but identification of MOG-Abs predicted a non-MS course at 1-year follow-up: only 2/23 (9%) MOG-Ab-positive patients were diagnosed with MS compared to 16/42 (38%) MOG-Ab-negative patients (p = 0.019, Fisher exact test). Antibody positivity at outset was a useful predictor for a non-MS disease course, with a positive predictive value of 91% (95% confidence interval [CI] 72-99), negative predictive value of 38% (95% CI 24-54), positive likelihood ratio of 4.02 (CI 1.0-15.4), and odds ratio of 6.5 (CI 1.3-31.3). CONCLUSIONS MOG-Abs are found at presentation in 35% of patients with childhood ADS, across a range of demyelinating disorders. Antibody positivity can be useful in predicting a non-MS disease course at onset.',\n","  3.621069192264962)]"]},"metadata":{},"execution_count":16}]},{"cell_type":"code","source":["qrels = []\n","for qrel in dataset.qrels_iter(): # reading in all the relevancy scores\n","    qrels.append(qrel)"],"metadata":{"id":"EH1lTrrXcKy0","executionInfo":{"status":"ok","timestamp":1647985434100,"user_tz":0,"elapsed":212,"user":{"displayName":"Alexander C","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"06292930002536064627"}}},"execution_count":17,"outputs":[]}]}